### ISSUE TITLE
macOS 12.2 Kernel Panic when Hibernating if Docker Is Running

### ISSUE THREAD

**Issue by jscheel** [docker/for-mac#6191]
- [x] I have tried with the latest version of Docker Desktop
  - [x] I have tried disabling enabled experimental features
  - [x] I have uploaded Diagnostics
  - Diagnostics ID: 78CD1D75-0E43-4692-9F0E-2BC7C2C574C4/20220222150901

### Expected behavior
macOS is able to go into full hibernation even if docker is running, without causing a kernel panic.

### Actual behavior
Kernel panic issued with the message `Sleep transition timed out after 180 seconds while creating hibernation file or while calling rootDomain's clients about upcoming rootDomain's state changes`. This only happens when docker is running. If I turn off docker, macOS goes into hibernation without issue.

### Information

  - macOS Version: 12.2
  - Intel chip or Apple chip: intel i9
  - Docker Desktop Version: 4.4.2 (73305)

This is a reproducible issue. It only panics if my computer goes to sleep when docker is running. Here is an additional issue that was closed due to inactivity: https://github.com/docker/for-mac/issues/5715

### Output of `/Applications/Docker.app/Contents/MacOS/com.docker.diagnose check`

```
[PASS] DD0027: is there available disk space on the host?
[PASS] DD0028: is there available VM disk space?
[PASS] DD0031: does the Docker API work?
[PASS] DD0004: is the Docker engine running?
[PASS] DD0011: are the LinuxKit services running?
[PASS] DD0016: is the LinuxKit VM running?
[PASS] DD0001: is the application running?
[PASS] DD0018: does the host support virtualization?
[PASS] DD0017: can a VM be started?
[PASS] DD0015: are the binary symlinks installed?
[PASS] DD0003: is the Docker CLI working?
[PASS] DD0013: is the $PATH ok?
[PASS] DD0007: is the backend responding?
[PASS] DD0014: are the backend processes running?
[PASS] DD0008: is the native API responding?
[PASS] DD0009: is the vpnkit API responding?
[PASS] DD0010: is the Docker API proxy responding?
[PASS] DD0012: is the VM networking working?
[PASS] DD0032: do Docker networks overlap with host IPs?
[SKIP] DD0030: is the image access management authorized?
[PASS] DD0019: is the com.docker.vmnetd process responding?
[PASS] DD0033: does the host have Internet access?
No fatal errors detected.
```

### Steps to reproduce the behavior

  1. Start docker and kubernetes (though this was happening before using kubes as well).
  2. Allow macOS to go into deep hibernate (default is 3 hours after sleep)

**Comment by jscheel** [comment]
Anyone else running into this? I can still replicate every time.

**Comment by jscheel** [comment]
My org is a paying org for docker, should I file this as a support request?

**Comment by ErikFeeley** [comment]
This happens to me on my work machine (MBP 16) However I think it only will panic if there is a container actually running when I close the lid and come back after enough times has passed to go into hibernate.

**Comment by myzticman** [comment]
Seems to be happening to me too. Didn't realise at first it may be docker related. Will try closing docker completely every time to see if it rectifies.

**Comment by FelipeUgarte** [comment]
I have the same problem: "Sleep transition timed out after 180 seconds while creating hibernation file or while calling rootDomain's clients about upcoming rootDomain's state changes"

I have found that if I have my hub plugged to the MacBook the problem happens, if I leave it disconnected the problem is gone.

This is my hub: https://www.hypershop.com/collections/hyperdrive/products/hyperdrive-gen2-12-in-1-usb-c-hub
I have send a mail to the hyperdrive support and I'm still waiting for a replay.

**Comment by klemmari1** [comment]
Same happens for me, Intel MBP16 2019. Running latest version of Docker Desktop (4.8.2) and macOS Monterey 12.4.

This only happens when I'm on battery power over night and Docker is running when I put it the MacBook to sleep. So if the MacBook is plugged into my monitor which feeds it power, it will not crash.

If I quit Docker Desktop first, put the laptop to sleep and let it be on battery power over night, it will not crash.

Pretty much only settings I've touched on Docker Desktop are some resource settings, giving it a bit more CPU and memory.

**Comment by squallsama** [comment]
> Same happens for me, Intel MBP16 2019. Running latest version of Docker Desktop (4.8.2) and macOS Monterey 12.4.
> 
> This only happens when I'm on battery power over night and Docker is running when I put it the MacBook to sleep. So if the MacBook is plugged into my monitor which feeds it power, it will not crash.
> 
> If I quit Docker Desktop first, put the laptop to sleep and let it be on battery power over night, it will not crash.
> 
> Pretty much only settings I've touched on Docker Desktop are some resource settings, giving it a bit more CPU and memory.

Yep. Same experience.

**Comment by mariaines** [comment]
Same issue for me. I actually sent in my MBP to Apple twice for servicing: the first time they replaced the logic board -- same issue; the second time they just wiped it, ran their tests, said everything works fine on a factory reset so it's not an Apple problem 😡 

Here's a previous issue in this repo from over a year ago: https://github.com/docker/for-mac/issues/5715. @djs55 [closed](https://github.com/docker/for-mac/issues/5715#issuecomment-892659472) it saying it must be an Apple problem.

Not sure what to do at this point. Whether it is a MacOS or hardware issue, clearly something Docker is doing is triggering it, so it's frustrating that no one on Docker team can prioritize looking into this / working with Apple directly to solve the problem.

Would love to hear if anyone has found a decent workaround.

**Comment by MarcMielke** [comment]
suffering the same on 12.4 with 4.9.1. These crashes do not occur if I don't leave docker running while on battery

**Comment by igin** [comment]
Same on macOS 12.4 with docker for mac 4.9.0.

**Comment by jscheel** [comment]
Paying account, filed several support requests but never heard back from Docker. It's been months. I'm not really sure what to do.

**Comment by nFec** [comment]
Same here in 4.9.1 on macOS 12.4

**Comment by MarcMielke** [comment]
@jscheel want to change the title? This looks like a generic, broader underlying issue, occurring throughout various combos of macos/ddfm versions. So far we've got reports for macos versions 12.2 - 12.4 and ddfm versions ranging from 4.4.2 - 4.9.1. It also seems to be Intel specific - no M1 reports so far.

**Comment by djs55** [comment]
A kernel crash indicates a bug in macOS (or the hardware) both of which are managed by Apple. I think the best thing to do is for everyone who experiences this to use the macOS feedback assistant to report the bug directly to Apple. Please post the bug report IDs here so we can make a collection of them.

**Comment by clintmod** [comment]
There's a thread here about disabling hibernate to mitigate: https://discussions.apple.com/thread/250908868

```
sudo pmset -a hibernatemode 0
```

**Comment by vgjenks** [comment]
> A kernel crash indicates a bug in macOS (or the hardware) both of which are managed by Apple. I think the best thing to do is for everyone who experiences this to use the macOS feedback assistant to report the bug directly to Apple. Please post the bug report IDs here so we can make a collection of them.

I'm actually working with them right now in an open ticket, for this very issue. I mentioned that I suspected Docker because I was fine until I upgraded to 4.9+ Somehow, now, going back to an older version doesn't help.

I'm sending daily crash reports and uploaded them to my ticket as well. Nothing yet. I don't hold out hope that Apple will do anything about this quickly, being a massive company with bigger fish to fry. I think we have to hope Docker figures out the cause and at least helps sort it out!

**Comment by jscheel** [comment]
> @jscheel want to change the title? This looks like a generic/broader underlying issue, affecting various combos of macos/ddfm versions. So far we've got reports for macos versions 12.2 - 12.4 and ddfm versions ranging from 4.4.2 - 4.9.1. It also seems to be Intel specific - no M1 reports so far.

Sure thing

**Comment by jscheel** [comment]
> There's a thread here about disabling hibernate to mitigate: https://discussions.apple.com/thread/250908868
> 
> ```
> sudo pmset -a hibernatemode 0
> ```

This isn't a good long term solution though. This means your laptop will drain its battery much quicker when asleep and not plugged in.

**Comment by vgjenks** [comment]
> > There's a thread here about disabling hibernate to mitigate: https://discussions.apple.com/thread/250908868
> > ```
> > sudo pmset -a hibernatemode 0
> > ```
> 
> This isn't a good long term solution though. This means your laptop will drain its battery much quicker when asleep and not plugged in.

Right. It disables a major, critical feature of the OS. Unfortunately, there's not another good option for Intel Mac folks. I lose about 1/3 of my battery overnight. A weekend w/o plugging in = drained battery.

I practically live in Docker and run so many containers (microservices on several different client projects), that it defeats the purpose of shutting Docker down every night so as to prevent a kernel panic. Whether I shut it down or it crashes on its own, it's a big distraction to get everything back into running order again.

**Comment by wwangsa** [comment]
This happened to me as well. I stopped a running container and closed the lid when I was done with work. It went to sleep then it crashed when I woke it up this morning. 

I'm on MBP16 2019, OS X 12.5, Docker desktop 4.5 (Engine 20.10.12, Credential Helper: 0.6.4, Snyk: v1.827.0, Compose: 1.29.2, Kubernetes: v1.22.5).

**Comment by vgjenks** [comment]
> This happened to me as well. I stopped a running container and closed the lid when I was done with work. It went to sleep then it crashed when I woke it up this morning.
> 
> I'm on MBP16 2019, OS X 12.5, Docker desktop 4.5 (Engine 20.10.12, Credential Helper: 0.6.4, Snyk: v1.827.0, Compose: 1.29.2, Kubernetes: v1.22.5).

I've resorted to closing all containers (dozens...ugh) and shutting down Docker Desktop, every night. Far from ideal, but at least I'm not doing a full reboot every morning and getting that much more distracted from my work.

Personally, I've had nothing but problems with Docker outside of a Linux desktop and servers. That's the only place none of this type of instability happened. I had so many problems on Windows, I switched to a Mac...and here we are. 🤷‍♂️

**Comment by clintmod** [comment]
to be clear… disabling hibernation via `sudo pmset -a hibernatemode 0` doesn't prevent sleep it just prevents dumping the ram to disk so that your laptop can still boot up to the same state in the event of complete battery failure

**Comment by marcotartagni** [comment]
I just got a resolutive solution for me. I think that most of the problems are due to kernel extensions. In my case, was not Docker but VirtualBox, so I tried:

# sudo kextcache --clear-staging

(maybe this is enough for Docker...) then I also reinforced the command with:

# sudo kextload -b org.virtualbox.kext.VBoxDrv
# sudo kextload -b org.virtualbox.kext.VBoxNetFlt
# sudo kextload -b org.virtualbox.kext.VBoxNetAdp
# sudo kextload -b org.virtualbox.kext.VBoxUSB

Don't ask me why, but it worked like a charm!

**Comment by mariaines** [comment]
@marcotartagni curious how you arrived at this solution? i.e. how did you identify / confirm it was a virtualbox kernel extension? Did this solve the problem forever, or do you have do this at some regular cadence? Basically just hoping to understand your theory a bit more before I type in a kernel command that I don't fully understand 😝

**Comment by marcotartagni** [comment]
> @marcotartagni curious how you arrived at this solution? i.e. how did you identify/confirm it was a virtualbox kernel extension? Did this solve the problem forever, or do you have to do this at some regular cadence? Basically just hoping to understand your theory a bit more before I type in a kernel command that I don't fully understand 😝

Yes, still working well, I did it once. Actually, I used it in also the case of error "Kernel Driver Not Installed (Rc=-1908)" that sometimes happens in upgrading OS on Mac; see 
https://forums.virtualbox.org/viewtopic.php?f=8&t=99948
https://forums.virtualbox.org/viewtopic.php?f=8&t=102603
Again, I do not know insights, do first a full backup, the usual jingle...

**Comment by konsalex** [comment]
Same issue here 👍

**Comment by klemensz** [comment]
> I think the best thing to do is for everyone who experiences this to use the macOS feedback assistant to report the bug directly to Apple.

I have done this a hundred times or so already over the last years ... 🤷‍♂️ I think Apple doesn't give a ****

**Comment by kerbylane** [comment]
I began experiencing this every time I closed my laptop immediately after increasing the RAM allocated to Docker from 8gb to 32gb, half of my laptop's RAM. Dropping it to 16gb resolved it most of the time.

**Comment by klemmari1** [comment]
> I began experiencing this every time I closed my laptop immediately after increasing the RAM allocated to Docker from 8gb to 32gb, half of my laptop's RAM. Dropping it to 16gb resolved it most of the time.

This actually fixed the issue for me as well 👍 Thanks for the suggestion!

**Comment by jscheel** [comment]
Interesting, I wonder if the ram allocation indicates this has something to do with creating the sleep image.

**Comment by wwangsa** [comment]
Since I upgraded to Ventura, the daily crash has been caused by "WindowsServer" being unresponsive. You can see this reported everywhere. The trigger for me is plugging/unplugging external monitors but sometimes it crashed randomly. I wonder if this is a macOS issue rather than Docker's.

**Comment by jscheel** [comment]
> 

I'll definitely give this a try, but I've had the issue happen on the same updates with fresh hibernation images.
